examples/server/utils.hpp

```diff
@@ -167,6 +167,34 @@ static T json_value(const json &body, const std::string &key, const T &default_v
         : default_value;
 }
 
+inline std::string format_llama2(std::vector<json> messages)
+{
+    std::ostringstream output;
+    bool is_inside_turn = false;
+
+    for (auto it = messages.begin(); it != messages.end(); ++it) {
+        if (!is_inside_turn) {
+            output << "[INST] ";
+        }
+        std::string role    = json_value(*it, "role", std::string("user"));
+        std::string content = json_value(*it, "content", std::string(""));
+        if (role == "system") {
+            output << "<<SYS>>\n" << content << "\n<<SYS>>\n\n";
+            is_inside_turn = true;
+        } else if (role == "user") {
+            output << content << " [/INST]";
+            is_inside_turn = true;
+        } else {
+            output << " " << content << " </s>";
```

Comment:
It's not clear to me whether EOS is actually supposed to be embedded in the history of the conversation (like chatml's im_end), or just included in the examples to show where the model's output ends.

Reply:
I can confirm that the EOS is suppose to be in the history. Using the code below, I obtained the output from the original tokenizer:

```python
from transformers import AutoTokenizer
tokenizer = AutoTokenizer.from_pretrained('mistralai/Mistral-7B-Instruct-v0.2')
history = [
    { 'role': 'user', 'content': 'hello' },
    { 'role': 'assistant', 'content': 'response' },
    { 'role': 'user', 'content': 'again' },
    { 'role': 'assistant', 'content': 'response' },
]
print(tokenizer.apply_chat_template(history, tokenize=False))
# <s>[INST] hello [/INST]response</s>[INST] again [/INST]response</s>
print(tokenizer.apply_chat_template(history, tokenize=True))
# [1, 733, 16289, 28793, 6312, 28709, 733, 28748, 16289, 28793, 4439, 2, 733, 16289, 28793, 1076, 733, 28748, 16289, 28793, 4439, 2]
```

The EOS token (mapped to 2) is placed between turns. `[INST]` is mapped to 733, 16289, 28793. (some models add `[INST]` as special, so only cost one token)

Also one thing to note is that only some models supports `<<SYS>>` for system prompt. If the model haven't been trained for that  `<<SYS>>` , it will be considered as part of the first message from user.

One thing not consistent among different variations of this template is the space character between [/INST] and the response. I add it just for sure (does not change anything on the output)